### ISSUE TITLE
implement pthread_setcancelstate() 

### ISSUE THREAD

**Issue by nyh** [cloudius-systems/osv#737]
When running some programs in OSv, we see an annoying (and somewhat scary) message about pthread_setcancelstate() and/or pthread_setcanceltype() being stubbed.

There is really no reason to warn about these functions being used: It's really trivial to fully implement them, by remembering their current state in a thread local variable - either a separate __thread or another field in class pthread in pthread.cc..

The harder part is to implement pthread_cancel() - but a lot of code seems to be using pthread_setcancelstate() without ever using pthread_cancel(), so it would be nice to get rid of the annoying message in that case.

**Comment by myechuri** [comment]
> running some programs in OSv

@nyh : can you please share which programs found the issue so that i can verify fix? If there is no unit test, happy to write one. Thanks!

**Comment by nyh** [comment]
The last report I had on those warnings were from a user who ran an OpenMPI use case, but this use case was quite hard to replicate so I don't suggest you try that. I thought this was common, but trying a few random packages in apps/ I don't see it happening, so it might be less common than I thought.

Anyway, reproducing this in a unit test should be trivial - call pthread_setcancelstate() twice and see that the state is remembered (you can create another thread to see it this setting is per thread). Of course, beyond remembering this setting, this function won't do anything because we are not implementing pthread_cancel() yet.

By the way, to write a unit test create a new program in tests/ (I suggest you copy one which works on both OSv and Linux - e.g., tst-chdir.cc - so you can verify that OSv's behavior is identical to Linux's). To allow your test to run on both OSv and Linux, use standard C++ stuff, like std::thread, etc., not OSv-specific APIs. To compile the test for OSv, add it to the "tests" list in modules/tests/Makefile, run "scripts/build image=tests", and the "scripts/run.py -e tests/tst-yourtest.so" to run only your new test.

**Comment by myechuri** [comment]
@nyh : Thanks a lot for sharing background information on the warning, detailed steps for test workflow and test placement! Will update the thread once the test is ready.

Once test is ready, will submit PR for just pthread_setcancelstate first, and then do a separate PR for pthread_setcanceltype so that the changes are easier to review. Thanks!

**Comment by nyh** [comment]
On Tue, May 10, 2016 at 7:29 AM, myechuri <notifications@github.com> wrote:

> Once test is ready, will submit PR for just pthread_setcancelstate first,
> and then do a separate PR for pthread_setcanceltype so that the changes are
> easier to review. Thanks!
>

Thanks. Please send patches to the mailing list instead of using Github
Pull Requests. With patches on the mailing list, it is easier for everyone
(and not just one specific reviewer) to follow the review process.

**Comment by myechuri** [comment]
> Please send patches to the mailing list instead of using Github Pull Requests. 

Will do. Thanks, @nyh !

**Comment by myechuri** [comment]
@nyh : emailed osv-dev patch for [the commit](https://github.com/myechuri/osv/commit/c0d25a9eb40d4e5f1e31dd01535b411411055454). Testing done is [linked out](https://github.com/myechuri/osv/commit/c0d25a9eb40d4e5f1e31dd01535b411411055454#commitcomment-17494791) in the patch as well. Thanks for your guidance!

**Comment by myechuri** [comment]
Status update: patch for pthread_setcancelstate is [committed](https://github.com/cloudius-systems/osv/commit/ca6bcd4128e13175879933c328281cf21bc21d57).

**Comment by nyh** [comment]
Next time, if you include in your patch message the text "Fixes #737", the issue will automatically be closed when the patch is committed. For now, I'll close this issue manually.

**Comment by myechuri** [comment]
>  if you include in your patch message the text "Fixes #737", the issue will automatically be closed when the patch is committed

Will do, thanks for the pointer.

> For now, I'll close this issue manually.

Thanks, @nyh !
